### DISCUSSION TITLE
Is minio object version ID unique?

### DISCUSSION THREAD

**Discussion by OlegYankovic** [minio/minio#20310] [Q&A] [answered]
The documentation said that object version ID is unique. But according to documentation uniqueness is achieved by the fact that version ID is big (128 bit) random value. Here is quote from documentation:

> MinIO generates a unique and immutable identifier for each versioned object as part of write operations. Each object version ID consists of a 128-bit fixed-size [UUIDv4](https://datatracker.ietf.org/doc/html/rfc4122.html#section-4.4). UUID generation is sufficiently random to ensure high likelihood of uniqueness for any environment, are computationally difficult to guess, and do not require centralized registration process and authority to guarantee uniqueness.

So, we can only hope that version ID will be unique with high likelihood or there is an internal undocumented mechanism to achieve pure uniqueness?

**Comment by klauspost** [comment, marked as answer]
@OlegYankovic As stated they are UUID v4, which means they are random. Unless you go crazy with versions on an object there is no real chance they will ever collide.

If my math is correct, with 10,000 versions, the chance of a (birthday) collision is 1 in 106348874550248294661770741897 with 122 bits of entropy.

**Reply by OlegYankovic** [reply]
Ok, thank you for good explanation.
 
Actually problem, I try to solve, is storing different files with the same filename in minio. Obviously I need possibility to download every file later. And the simplest solution seems to be using the filename as object name and associate every file with filename+versionID. According to your explanation this will reduce the collisions nearly to zero. But if I want to go crazy and completely eliminate collisions, how I can do it regarding my task? For example, is it possible to configure minio to return errors on overwriting object?

**Reply by klauspost** [reply]
You will be fine as you keep your version count reasonable - you should never hit 10K versions.  If it was a global UUID across billions of objects, I could maybe be slightly concerned.
